File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
"""
Time = O(n)
Space = O(n)
"""
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if not s:
            return 0
        left = 0
        longest = 1
        char_set = set()
        char_set.add(s[left])
        
        for right in range(1, len(s)):
            if s[right] not in char_set:
                char_set.add(s[right])
                longest = max(longest, right - left + 1)
            else:
                while s[left] != s[right]:
                    if s[left] in char_set:
                        char_set.remove(s[left])
                    left += 1
                left += 1
        
        return longest
            

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int res = 0;
        unordered_map<char, int> charMap;
        
        int left = 0;
        for (int right = 0; right < s.size(); right++) {
            charMap[s[right]]++;
            if (charMap[s[right]] > 1) {
                while (charMap[s[right]] > 1) {
                    charMap[s[left]]--;
                    if (charMap[s[left]] == 0) {
                        charMap.erase(s[left]);
                    }
                    left++;
                }
            }
            res = max(res, right - left + 1);
        }

        return res;
    }
};